Add action to clear all buffers for a specific pane

alekspickle opened the pull request:
Fixes #2081 (?)

I want to start a bigger feature contribution but I needed some easy task to get familiar with the codebase. This seemed like a perfect candidate.

It's just a draft anyway if for some reason this naive implementation is somehow not favorable.
Also I guess for the sake of fixing this particular issue, it makes sense for me to clear only `line_above`.
But this is the more accurate semantic for this action IMO. Let me know what you guys think.

Tested it locallly , worked as expected.
- [x] test manually clear action (`zellij action clear` for a trashed pane)
- [x] write a dedicated test for it (not sure I did it clean here)

imsnif commented on src/sessions.rs:
```diff
@@ -207,7 +207,7 @@ pub(crate) fn assert_session_ne(name: &str) {
     }
     if name.contains('/') {
-        eprintln!("Session name cannot contains '/'.");
+        eprintln!("Session name cannot contain '/'.");
```
<3

imsnif commented on zellij-server/src/panes/grid.rs:
```diff
@@ -1109,7 +1109,13 @@ impl Grid {
+    pub fn clear_screen(&mut self) {
+        self.lines_above = Default::default();
+        self.lines_below = Default::default();
+        self.viewport = Default::default();
```
Let's add a conditional here so that we don't do this if `alternate_screen` is `Some`. This happens when eg. programs running in the terminal switch to full screen (eg. `htop` and friends - also Zellij itself). I think in such cases we really don't want to do this. We can add a log if that happens.

imsnif commented on zellij-server/src/tab/mod.rs:
```diff
@@ -235,6 +235,9 @@ pub trait Pane {
     fn pull_up(&mut self, count: usize);
+    fn clear_screen(&mut self) {
+        todo!();
+    }
```
Why not forgo the default implementation here? If you're concerned about plugins (our only other type of `Pane`) then I think we can leave it and remove the `todo!()` - I don't think this action is relevant for plugins...

alekspickle replied:
Not really when I was doing this specifically I was trying to understand the whole architecture of zellij, I think I can leave all the default stuff here without conditional you mentioned

imsnif commented on zellij-server/src/tab/unit/tab_integration_tests.rs:
```diff
@@ -708,6 +708,35 @@ fn dump_screen() {
+        "",
+        "screen was cleared properly"
+    );
+}
```
Nice!

imsnif reviewed (changes requested):
Hey, looks cool - good job on this!

I left some comments in the code. Otherwise let's also add a bindable key for this, you can take a look at similar implementations in our kdl parser: https://github.com/zellij-org/zellij/blob/main/zellij-utils/src/kdl/mod.rs

alekspickle commented:
@imsnif  Hmm, so `zj action clear`  works as expected, but the binded action is glitchy, don't know why. Could it be overlapping with something?
![zj-buggy](https://user-images.githubusercontent.com/22867443/225428193-47fbfbee-dd07-485b-bff9-5f631023aba0.gif)

jurica commented:
> Hmm, so `zj action clear` works as expected, but the binded action is glitchy, don't know why. Could it be overlapping with something? ![zj-buggy](https://user-images.githubusercontent.com/22867443/225428193-47fbfbee-dd07-485b-bff9-5f631023aba0.gif) [ ![zj-buggy](https://user-images.githubusercontent.com/22867443/225428193-47fbfbee-dd07-485b-bff9-5f631023aba0.gif) ](https://user-images.githubusercontent.com/22867443/225428193-47fbfbee-dd07-485b-bff9-5f631023aba0.gif) [ ](https://user-images.githubusercontent.com/22867443/225428193-47fbfbee-dd07-485b-bff9-5f631023aba0.gif)
> 
> I found [this comment](https://github.com/alacritty/alacritty/issues/1811#issuecomment-439719184) about escape sequences, but this is not my strongest side, so is there already some og this escape magic somewhere in zellij? `[Unhandled CSI] action='l', args=[2004], intermediates=[63] -> Reset bracketed paste `
> 
> At a glance looks like a limitation of sorts

What a coincidence, I was testing your branch tonight, just while you did your latest changes.
I also noticed that triggering the action via CLI behaves different than when binding it to a key. But I just realized that in a release build it's less glitchy... I hope that hint helps.

Apart from that glitch on keybinding, this is really awesome! For me this is the feature I missed the most in zellij, thanks for the implementation!

alekspickle commented:
Ok here are the logs I've localized near the buggy behavior:

<details>
<summary>logs</summary>
```

WARN   |zellij_server::panes::gri| 2023-03-21 11:14:42.109 [screen    ] [zellij-server/src/panes/grid.rs:2990]: Unhandled esc_dispatch: 62->[]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:42.110 [screen    ] [zellij-server/src/panes/grid.rs:2382]: Unhandled osc: [[49], [116, 114, 101, 101]]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:42.398 [screen    ] [zellij-server/src/panes/grid.rs:2382]: Unhandled osc: [[49], [46, 46, 111, 99, 117, 109, 101, 110, 116, 115, 47, 114, 117, 115, 116]]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:42.398 [screen    ] [zellij-server/src/panes/grid.rs:2990]: Unhandled esc_dispatch: 61->[]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:50.162 [screen    ] [zellij-server/src/panes/grid.rs:2990]: Unhandled esc_dispatch: 62->[]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:50.163 [screen    ] [zellij-server/src/panes/grid.rs:2382]: Unhandled osc: [[49], [46, 46, 111, 99, 117, 109, 101, 110, 116, 115, 47, 114, 117, 115, 116]]
WARN   |zellij_server::panes::gri| 2023-03-21 11:14:50.174 [screen    ] [zellij-server/src/panes/grid.rs:2990]: Unhandled esc_dispatch: 61->[]

```
</details>

So I need to handle it in esc_dispatch somehow, I have no idea what to do here.

imsnif commented:
Hey @alekspickle - sorry it took me some time to get to this. Thank you for your patience.

I have not managed to reproduce the glitch you mention, but on the other hand this as it is now doesn't really work for me locally with various results. I have had the most success with this change (in grid.rs):
```rust
    pub fn clear_screen(&mut self) {
        if self.alternate_screen_state.is_some() {
            log::warn!("Tried to clear pane with alternate_screen_state");
            return;
        }
        self.reset_terminal_state();
        self.mark_for_rerender();
//         self.lines_above = Default::default();
//         self.lines_below = Default::default();
//         self.viewport = Default::default();
    }

```

What do you think?

alekspickle commented:
I was able to play with a several existing functions, but the best  I did is that the new prompt line was at the bottom and after you hit enter same you have. This is better, and honestly - good enough. My idea was to clean up for a really trashed panes and I would not really override it by default.

imsnif commented:
Cool, glad you like this variant. Would you like to adjust the PR?

Otherwise, for the binding being messed up - I do not see this behaviour (neither with the change nor with the original PR).  Do you still see it with the change?

alekspickle commented:
I already did!
Yeah, it's kinda weird, but reasonable enough:
![zj-release](https://user-images.githubusercontent.com/22867443/228495148-32ef33c6-2b88-4202-a2fe-4408a7f46e57.gif)

alekspickle commented:
also I have new assets to commit after building release version for some reason:
```

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   zellij-utils/assets/plugins/compact-bar.wasm
        modified:   zellij-utils/assets/plugins/status-bar.wasm
        modified:   zellij-utils/assets/plugins/strider.wasm
        modified:   zellij-utils/assets/plugins/tab-bar.wasm

```
